File: src/Nerdbank.GitVersioning/CloudBuildServices/Travis.cs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Nerdbank.GitVersioning.CloudBuildServices;

/// <summary>
/// Travis CI build support.
/// </summary>
/// <remarks>
/// The Travis CI environment variables referenced here are <see href="https://docs.travis-ci.com/user/environment-variables/#default-environment-variables">documented here</see>.
/// </remarks>
internal class Travis : ICloudBuild
{
    // TRAVIS_BRANCH can reference a branch or a tag, so make sure it starts with refs/heads

    /// <inheritdoc/>
    public string BuildingBranch => CloudBuild.ShouldStartWith(Environment.GetEnvironmentVariable("TRAVIS_BRANCH"), "refs/heads/");

    /// <inheritdoc/>
    public string BuildingTag => Environment.GetEnvironmentVariable("TRAVIS_TAG");

    /// <inheritdoc/>
    public string GitCommitId => Environment.GetEnvironmentVariable("TRAVIS_COMMIT");

    /// <inheritdoc/>
    public bool IsApplicable => !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("TRAVIS"));

    /// <inheritdoc/>
    public bool IsPullRequest => !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("TRAVIS_PULL_REQUEST_BRANCH"));

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildNumber(string buildNumber, TextWriter stdout, TextWriter stderr)
    {
        return new Dictionary<string, string>();
    }

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildVariable(string name, string value, TextWriter stdout, TextWriter stderr)
    {
        return new Dictionary<string, string>();
    }
}

File: src/nerdbank-gitversioning-rs/src/cloud_build_services/travis.rs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

//! [Travis CI](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables) support.

use std::io::{self, Write};

use crate::cloud::{
    CloudBuild, EnvironmentVariables, empty_variables, env, is_nonempty, should_start_with,
};

/// Travis CI cloud-build integration.
pub struct Travis;

impl CloudBuild for Travis {
    fn name(&self) -> &'static str {
        "Travis"
    }

    fn is_applicable(&self) -> bool {
        is_nonempty(env("TRAVIS").as_deref())
    }

    fn is_pull_request(&self) -> bool {
        is_nonempty(env("TRAVIS_PULL_REQUEST_BRANCH").as_deref())
    }

    fn building_branch(&self) -> Option<String> {
        should_start_with(env("TRAVIS_BRANCH"), "refs/heads/")
    }

    fn building_tag(&self) -> Option<String> {
        env("TRAVIS_TAG")
    }

    fn git_commit_id(&self) -> Option<String> {
        env("TRAVIS_COMMIT")
    }

    fn set_cloud_build_number(
        &self,
        _build_number: &str,
        _stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        Ok(empty_variables())
    }

    fn set_cloud_build_variable(
        &self,
        _name: &str,
        _value: &str,
        _stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        Ok(empty_variables())
    }
}
